ambry-store/src/main/java/com/github/ambry/store/StorageManager.java

```diff
@@ -259,6 +256,49 @@ void verifyDiskHealth() throws StoreException {
     }
   }
 
+  /**
+   * Maybe restore disk's availability and update instance capacity.
+   * @return A set of {@link DiskId}s that are still unavailable.
+   */
+  Set<DiskId> maybeRestoreDiskAvailability() {
+    List<DiskId> allDiskIds = currentNode.getDiskIds();
+    Set<DiskId> failedDiskIds = allDiskIds.stream()
+        .filter(diskId -> diskId.getState() == HardwareState.UNAVAILABLE)
+        .collect(Collectors.toSet());
+    if (!storeConfig.storeRestoreUnavailableDiskInFullAuto || !clusterMap.isDataNodeInFullAutoMode(currentNode)) {
+      return failedDiskIds;
+    }
+    // We should try to restore the availability of failed disks
+    if (!failedDiskIds.isEmpty()) {
+      List<DiskId> disksToRecover = new ArrayList<>();
+      for (DiskId diskId : failedDiskIds) {
+        // If the mount path exist and is a directory, it's likely disk is already recovered
+        File mount = new File(diskId.getMountPath());
```

Comment:
You have some explanation in the PR title but I don't fully understand.
Let's say one partition is corrupted and we marked it as failure in the Helix. Do we remove the folder right away?
Can the mount point existing is just the old mount point we haven't removed yet? How do we know it's the new one?

Reply:
This is not for one partition, this is for the disk. If a disk failed, all replicas on this disk would be moved to a different host/disk in DiskFailureHandler. And in DiskFailureHandler we would also mark this disk as "UNAVALABLE" in property store and decrease the instance disk capacity. 

Now if this disk is fixed, (or replaced), then we want to mark this disk as "AVAILABLE" in property store and increase the instance disk capacity when starting up storage manager so we don't have to manually do this every time.

But how do we know if the disk is fixed? One simple way is to check if the disk mount path exists or not. Because when a disk is broken, its mount path doesn't exist in the root filesystem. And now if it exists, it's a very strong evidence that this disk is good now.

This is not a perfect solution, but based on what I observed, it solves most of the time.